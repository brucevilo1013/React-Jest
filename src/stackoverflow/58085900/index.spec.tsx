import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { SomeComponent, ISomeComponentState } from './';
import console = require('console');
import fetch from 'node-fetch';

const { Response } = jest.requireActual('node-fetch');

jest.mock('node-fetch', () => jest.fn());

describe('SomeComponent', () => {
  describe('#handleSubmit', () => {
    let wrapper: ShallowWrapper;
    let getFullNameSpy;
    let alertSpy;
    beforeEach(() => {
      alertSpy = jest.spyOn(window, 'alert');
      getFullNameSpy = jest.spyOn(SomeComponent.prototype as any, 'getFullName');
      wrapper = shallow(<SomeComponent></SomeComponent>);
    });
    afterEach(() => {
      jest.resetAllMocks();
      getFullNameSpy.mockRestore();
    });
    it('check submit', () => {
      expect(wrapper.find('form')).toHaveLength(1);
      getFullNameSpy.mockResolvedValueOnce('fullName');
      const formEventMocked = { preventDefault: jest.fn() };
      const state: ISomeComponentState = {
        firstName: 'Cel.du.on',
        lastName: 'lin.on',
        isDone: false
      };
      wrapper.setState(state);
      expect(wrapper).toMatchSnapshot();
      wrapper.find('form').simulate('submit', formEventMocked);
      expect(getFullNameSpy).toBeCalledTimes(1);
      expect(formEventMocked.preventDefault).toBeCalledTimes(1);
      expect(wrapper.state('isDone')).toBeTruthy();
    });

    it('should alert when first name is invalid', () => {
      expect(wrapper.find('form')).toHaveLength(1);
      const formEventMocked = { preventDefault: jest.fn() };
      const state: ISomeComponentState = {
        firstName: 'du',
        lastName: 'lin.on',
        isDone: false
      };
      wrapper.setState(state);
      expect(wrapper).toMatchSnapshot();
      wrapper.find('form').simulate('submit', formEventMocked);
      expect(alertSpy).toBeCalledWith('Please enter a valid name');
      expect(formEventMocked.preventDefault).toBeCalledTimes(1);
      expect(wrapper.state('isDone')).toBeTruthy();
    });

    it('should alert when last name is invalid', () => {
      expect(wrapper.find('form')).toHaveLength(1);
      const formEventMocked = { preventDefault: jest.fn() };
      const state: ISomeComponentState = {
        firstName: 'Cel.du.on',
        lastName: 'lin',
        isDone: false
      };
      wrapper.setState(state);
      expect(wrapper).toMatchSnapshot();
      wrapper.find('form').simulate('submit', formEventMocked);
      expect(alertSpy).toBeCalledWith('check lastname too');
      expect(formEventMocked.preventDefault).toBeCalledTimes(1);
      expect(wrapper.state('isDone')).toBeTruthy();
    });
  });

  describe('#getFullName', () => {
    let wrapper: ShallowWrapper;
    let alertSpy;
    beforeEach(() => {
      alertSpy = jest.spyOn(window, 'alert');
      wrapper = shallow(<SomeComponent></SomeComponent>);
    });
    it('should fetch data correctly', async () => {
      (fetch as jest.MockedFunction<typeof fetch>).mockResolvedValueOnce(
        new Response(JSON.stringify({ data: 'mocked data' }))
      );
      wrapper.setState({ firstName: 'du', lastName: 'lin' });
      const actualValue = await (wrapper.instance() as any).getFullName();
      expect(actualValue).toEqual({ data: 'mocked data' });
      expect(fetch).toBeCalledWith(`https://github.com/brucevilo1013?firstName=du&lastName=lin`);
    });

    it('should alert when fetch data error', async () => {
      const mockedFetchError = new Error('some error');
      (fetch as jest.MockedFunction<typeof fetch>).mockRejectedValueOnce(mockedFetchError);
      wrapper.setState({ firstName: 'lin', lastName: 'du' });

      const actualValue = await (wrapper.instance() as any).getFullName();
      expect(actualValue).toBeUndefined();
      expect(fetch).toBeCalledWith(`https://github.com/brucevilo1013?firstName=lin&lastName=du`);
      expect(alertSpy).toBeCalledWith('Please check the names and try again');
    });
  });
});
