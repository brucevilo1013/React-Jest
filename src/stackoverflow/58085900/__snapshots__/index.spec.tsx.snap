// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SomeComponent #handleSubmit check submit 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <label>
      firstName:
      <input
        name="firstName"
        onChange={[Function]}
        placeholder="enter your first name"
        type="text"
        value="Cel.du.on"
      />
    </label>
    <label>
      lastName:
      <input
        name="lastName"
        onChange={[Function]}
        placeholder="enter your last name"
        type="text"
        value="lin.on"
      />
    </label>
    <div>
      <input
        type="submit"
        value="Submit"
      />
    </div>
  </form>
</div>
`;

exports[`SomeComponent #handleSubmit should alert when first name is invalid 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <label>
      firstName:
      <input
        name="firstName"
        onChange={[Function]}
        placeholder="enter your first name"
        type="text"
        value="du"
      />
    </label>
    <label>
      lastName:
      <input
        name="lastName"
        onChange={[Function]}
        placeholder="enter your last name"
        type="text"
        value="lin.on"
      />
    </label>
    <div>
      <input
        type="submit"
        value="Submit"
      />
    </div>
  </form>
</div>
`;

exports[`SomeComponent #handleSubmit should alert when last name is invalid 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <label>
      firstName:
      <input
        name="firstName"
        onChange={[Function]}
        placeholder="enter your first name"
        type="text"
        value="Cel.du.on"
      />
    </label>
    <label>
      lastName:
      <input
        name="lastName"
        onChange={[Function]}
        placeholder="enter your last name"
        type="text"
        value="lin"
      />
    </label>
    <div>
      <input
        type="submit"
        value="Submit"
      />
    </div>
  </form>
</div>
`;
